File: backend/recovery/hash_verifier.py
import hashlib
import os
from typing import Dict, Union, List

class HashVerifier:
    """
    Forensic Hash Verification Module.
    Responsible for calculating MD5, SHA-1, and SHA-256 checksums of files,
    directories, or binary streams to verify integrity and maintain a strict chain of custody.
    """

    @staticmethod
    def calculate_file_hashes(file_path: str, algorithms: List[str] = None) -> Dict[str, str]:
        """
        Calculate cryptographic hashes for a given file.
        """
        if not algorithms:
            algorithms = ["md5", "sha1", "sha256"]
        
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"File not found: {file_path}")
            
        hash_objects = {}
        for algo in algorithms:
            algo_lower = algo.lower()
            if algo_lower == "md5":
                hash_objects["md5"] = hashlib.md5()
            elif algo_lower == "sha1":
                hash_objects["sha1"] = hashlib.sha1()
            elif algo_lower == "sha256":
                hash_objects["sha256"] = hashlib.sha256()
            else:
                raise ValueError(f"Unsupported hashing algorithm: {algo}")

        # Read in blocks of 64KB to handle large forensic images efficiently
        block_size = 65536
        with open(file_path, "rb") as f:
            for block in iter(lambda: f.read(block_size), b""):
                for obj in hash_objects.values():
                    obj.update(block)

        return {algo: obj.hexdigest() for algo, obj in hash_objects.items()}

    @staticmethod
    def calculate_bytes_hashes(data: bytes, algorithms: List[str] = None) -> Dict[str, str]:
        """
        Calculate cryptographic hashes for a byte buffer.
        """
        if not algorithms:
            algorithms = ["md5", "sha1", "sha256"]
            
        results = {}
        for algo in algorithms:
            algo_lower = algo.lower()
            if algo_lower == "md5":
                results["md5"] = hashlib.md5(data).hexdigest()
            elif algo_lower == "sha1":
                results["sha1"] = hashlib.sha1(data).hexdigest()
            elif algo_lower == "sha256":
                results["sha256"] = hashlib.sha256(data).hexdigest()
            else:
                raise ValueError(f"Unsupported hashing algorithm: {algo}")
        return results

    @classmethod
    def verify_integrity(cls, file_path: str, expected_hash: str, algorithm: str = "sha256") -> bool:
        """
        Verify if the hash of a file matches the expected hash.
        """
        try:
            hashes = cls.calculate_file_hashes(file_path, [algorithm])
            calculated = hashes.get(algorithm.lower())
            return calculated.lower() == expected_hash.lower()
        except Exception:
            return False

File: src/backend/src/recovery/hash_verifier.ts
import crypto from "crypto";
import fs from "fs";

export class HashVerifier {
  /**
   * Calculate cryptographic hashes for a given file.
   */
  static calculateFileHashes(filePath: string, algorithms: string[] = ["md5", "sha1", "sha256"]): Record<string, string> {
    if (!fs.existsSync(filePath)) {
      throw new Error(`File not found: ${filePath}`);
    }

    const hashObjects: Record<string, crypto.Hash> = {};
    for (const algo of algorithms) {
      const algoLower = algo.toLowerCase();
      if (["md5", "sha1", "sha256"].includes(algoLower)) {
        hashObjects[algoLower] = crypto.createHash(algoLower);
      } else {
        throw new Error(`Unsupported hashing algorithm: ${algo}`);
      }
    }

    const fd = fs.openSync(filePath, "r");
    const buffer = Buffer.alloc(65536);
    let bytesRead = 0;
    let position = 0;

    while (true) {
      bytesRead = fs.readSync(fd, buffer, 0, buffer.length, position);
      if (bytesRead === 0) {
        break;
      }
      position += bytesRead;
      const dataChunk = buffer.subarray(0, bytesRead);
      for (const obj of Object.values(hashObjects)) {
        obj.update(dataChunk);
      }
    }
    fs.closeSync(fd);

    const results: Record<string, string> = {};
    for (const [algo, obj] of Object.entries(hashObjects)) {
      results[algo] = obj.digest("hex");
    }
    return results;
  }

  /**
   * Calculate cryptographic hashes for a byte buffer.
   */
  static calculateBytesHashes(data: Buffer | string, algorithms: string[] = ["md5", "sha1", "sha256"]): Record<string, string> {
    const buffer = typeof data === "string" ? Buffer.from(data, "utf8") : data;
    const results: Record<string, string> = {};

    for (const algo of algorithms) {
      const algoLower = algo.toLowerCase();
      if (["md5", "sha1", "sha256"].includes(algoLower)) {
        results[algoLower] = crypto.createHash(algoLower).update(buffer).digest("hex");
      } else {
        throw new Error(`Unsupported hashing algorithm: ${algo}`);
      }
    }
    return results;
  }

  /**
   * Verify if the hash of a file matches the expected hash.
   */
  static verifyIntegrity(filePath: string, expectedHash: string, algorithm = "sha256"): boolean {
    try {
      const hashes = this.calculateFileHashes(filePath, [algorithm]);
      const calculated = hashes[algorithm.toLowerCase()];
      return calculated.toLowerCase() === expectedHash.toLowerCase();
    } catch (e) {
      return false;
    }
  }
}
